### PR TITLE
 Introduces new "env_var" and "file" fields to Secret to allow specifying name/mountPath on injection #minor

### PR DESCRIPTION
**Pull request by gpgn**
# TL;DR
Introduces new fields to the `Secret` object:
- `env_var`
- `file`

Allowing users to directly specify a name or mountPath for the Secret, without having to specify a full PodTemplate(name). The old `mount_requirement` can still be used. Example:

```python

from flytekit import task, workflow, Secret
import flytekit
import os


@task(
    secret_requests=[
        Secret(
            group="user-info",
            key="user_secret",
            mount_requirement=Secret.MountType.FILE,
        ),
        Secret(
            group="user-info",
            key="user_secret",
            mount_requirement=Secret.MountType.ENV_VAR,
        ),
    ]
)
def old() -> None:
    context = flytekit.current_context()
    print("old")

    # mount_requirement=ENV_VAR
    print(context.secrets.get(env_name="_FSEC_USER-INFO_USER_SECRET"))

    # mount_requirement=FILE
    with open('/etc/flyte/secrets/user-info/user_secret', 'r') as infile:
        print(infile.readlines())
    return True

@task(
    secret_requests=[
        Secret(
            group="user-info",
            key="user_secret",
            env_var=Secret.MountEnvVar(name="foo"),
        ),
        Secret(
            group="user-info",
            key="user_secret",
            file=Secret.MountFile(path="/foo"),
        ),
    ]
)

def new() -> None:
    context = flytekit.current_context()
    print("new")

    # env_var=
    print(context.secrets.get(env_name="foo"))

    # file=
    with open('/foo/user_secret', 'r') as infile:
        print(infile.readlines())


@workflow
def training_workflow(hyperparameters: dict) -> None:
    """Put all of the steps together into a single workflow."""
    old()
    new()

```

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue

fixes https://github.com/flyteorg/flyte/issues/3053

## Follow-up issue
_NA_

## Linked PRs
- https://github.com/flyteorg/flytekit/pull/1726
- https://github.com/flyteorg/flytepropeller/pull/589
- https://github.com/flyteorg/flytesnacks/pull/1017

### CODE REVIEW

**Review by pingsutw — COMMENTED** [review]
LGTM, could you resolve a merge conflict

File: protos/flyteidl/core/security.proto [hunk 1]
```diff
@@ -42,7 +42,27 @@ message Secret {
...
     // +optional
-    MountType mount_requirement = 4;
+    MountType mount_requirement = 4 [deprecated=true];
+
+    // mount_target is optional. Indicates how the secret should be mounted. Secrets can be mounted as file, or as 
+    // environment variable. 
+    // +optional
+    oneof mount_target {
+        MountEnvVar env_var = 5;
+        MountFile file = 6;
```
Comment (wild-endeavor) [inline protos/flyteidl/core/security.proto:52]: since the existing `MountType` enum was only used in the `mount_requirement` field, it'd be effectively deprecated as well right?  Can we add the flag there as well?

Also since the existing enum had an Any concept, should we add that as well to the one of?

**Review by wild-endeavor — COMMENTED** [review]
This is great.  I want to call out though that this change needs to be both forward and backwards compatible.  That is,
* people on newer versions of the flyte backend (like upgrade propeller/admin/etc after these PRs are merged) but old versions of flytekit should continue to work, and
* people on existing versions of flyte backend (those who do not upgrade propeller/admin/etc with these PRs) but who do upgrade flytekit, should continue to work.
*

File: protos/flyteidl/core/security.proto (hunk 1 above)
Comment (wild-endeavor) [reply protos/flyteidl/core/security.proto:52]: Nm - let's ignore this comment

File: protos/flyteidl/core/security.proto [hunk 2]
```diff
@@ -42,7 +42,27 @@ message Secret {
...
+    }
+
+    // The name of the environment variable if the Secret is injected as environment variable. If ommitted, the default
```
Comment (wild-endeavor) [inline protos/flyteidl/core/security.proto:55]:
```suggestion
    // The name of the environment variable if the Secret is injected as environment variable. If
```

File: protos/flyteidl/core/security.proto [hunk 3]
```diff
@@ -42,7 +42,27 @@ message Secret {
...
+    // FLYTE_SECRETS_ENV_PREFIX prefix will be used.
```
Comment (wild-endeavor) [inline protos/flyteidl/core/security.proto:56]:
```suggestion
    // MountEnvVar is supplied with an empty string, FLYTE_SECRETS_ENV_PREFIX prefix will be used.
```

File: protos/flyteidl/core/security.proto [hunk 4]
```diff
@@ -42,7 +42,27 @@ message Secret {
...
+    message MountEnvVar {
+        string name = 1;
+    }
+
+    // The path where the Secret will be mounted. The execution will fail if the underlying key management system cannot 
+    // satisfy that requirement. If not provided, the default location will depend on the key management system.
```
Comment (wild-endeavor) [inline protos/flyteidl/core/security.proto:62]:
```suggestion
    // satisfy that requirement. If provided with an empty string, the default location will depend on the key management system.
```
